### ISSUE TITLE
Spaces clobbered in Firefox

### ISSUE THREAD

**Issue by madeleineostoja** [SimplaElements/simpla-richtext-behavior#23]
Space characters seem to be clobbered in Firefox as you type, and it's impossible to make a space between words unless you do a bunch of backspacing and messing about.

This is live in simpla-text as well, and is badly breaking.

Screencap repro:

![space clobbered](https://cloud.githubusercontent.com/assets/6374876/26548055/6607189e-4472-11e7-9fa9-faef3ce29b7f.gif)

**Comment by bedeoverend** [comment]
Seems like this is to do with CSS's `white-space` property. As per [MDN docs](https://developer.mozilla.org/en/docs/Web/CSS/white-space) by default white space characters will be collapsed. In contenteditable Chrome & Safari insert `&nbsp;` characters instead of spaces which stops the collapse, but in firefox it's regular spaces which get collapsed.

Fix for this is using `white-space: pre-wrap` which is what's used in ProseMirror's [default stylesheet](https://github.com/ProseMirror/prosemirror-view/blob/master/style/prosemirror.css#L7). Given RichText doesn't touch styles, this has to be applied by the elements i.e. `simpla-article` and `simpla-text`. 

Alternatively we could add inline styles to the given element, but that would obviously break specificity. I don't think we could use the injected style sheets hack like placeholder as the styles need to apply to the host when its an inline editor i.e. no wrapping. @seaneking any ideas?

**Comment by bedeoverend** [comment]
Fixed in simpla-text in https://github.com/SimplaElements/simpla-text/commit/14f33b5c34f4db86ac547dcadbe5e9b611333b41
Fixed in simpla-article in https://github.com/SimplaElements/simpla-article/commit/8f0b0f60ce5addfa77468c347ca22a4cb125ee5f

Doesn't fix the bug here, unless we decide we won't fix at this level, but I'll remove the breaking label as it can be patched up on the elements.

**Comment by madeleineostoja** [comment]
I'd say inline styles for thia because you specifically wouldn't want this overwritten, if it breaks core functionality like typing text without it

**Comment by madeleineostoja** [comment]
Or, probably cleaner, why not add to injected stylesheet, and then if inline just set inline tyke on host?

**Comment by madeleineostoja** [comment]
Should only have one injected stylesheet for everything though, IMO, maybe as a separate plugin.

I don't understand what the li rules are for in that default sheet, but probably want them.

**Comment by madeleineostoja** [comment]
Can also move some of the resets from article to the injected sheet.

Though ideally we should only include the styles that are needed, based on plugins available. Which is easy enough, just have CSS strings in JS per plugin (including placeholder) and then write to style tag based on what's available, with extra rule for inline whitespace rule for inline editors.

That seems a pretty nice solution to me, what say you @bedeoverend

**Comment by bedeoverend** [comment]
tl;dr Yeah I'm happy to group all plugin styles into some injected stylesheet

---

Yeah happy to go with an injected stylesheet. My only concern there would be:

> if inline just set inline tyke on host?

Someone having to apply different specificity (i.e. an `!important` for inline ones) for their styles based on whether the element is inline or not feels pretty confusing, for what is essentially just us cleaning up our implementation.

Then again, that really is a massive edge case at that point, you're talking about inline editors that want a custom `white-space` value.

Still, just getting the element to apply the styles feels a lot cleaner. Is there some way this can cleanly export a stylesheet of styles that the consumer can apply, or would that be a nightmare?

**Comment by madeleineostoja** [comment]
It's not an opinionated style, it's fixing an implementation bug. There is no reason to override whitespace on a simpla-text element. Some of the others, like break-word, sure. But whitespace just fixes cross browser bugs and it's a good thing it's forced.

Getting the elements to patch an implementation bug in RichText feels much messier to me. Doesn't matter that it happens to be CSS, it's not a stylistic thing, it's a bug.

**Comment by bedeoverend** [comment]
Yup that's fair. Lets move to integrating all of these styles in the behaviour, rather than asking for the elements to do it. Will aim for an injected stylesheet, but fallback to inline.

**Comment by madeleineostoja** [comment]
Adding `white-space: pre-wrap` to block level host elements added a bad regression with the box model, so now all simpla-text (and simpla-article) have mega spacing that you can't get rid of, breaking any layout they were used in. I'll roll back and publish a new version for simpla-text.

**Comment by madeleineostoja** [comment]
And that makes this a breaking bug again

**Comment by bedeoverend** [comment]
Now I see what you're talking about, I did notice the big padding but didn't realise what it meant / that it was a regression. Sorry - should have realised and caught that before pushing.

So given that `white-space: pre-wrap` (or `pre` the only two that preserve whitespace) cause this nasty bug, we can't use a CSS fix - checked and it doesn't matter how it's applied.

**Comment by bedeoverend** [comment]
Ok so the issue was the Shadow Root had spaces either side of the `<content></content>` tags, which added the extra spaces. Given what we want to do here, it'll probably take some time, so I'll re-apply the pre-wrap styles to text / article, but remove the whitespace around `<content></content>`

**Comment by bedeoverend** [comment]
So I attempted a purely JS solution to this - Essentially just replacing trailing spaces with non-breaking spaces (behavior in Chrome / Safari) whenever there's text input - but it ended up with way too many edge cases. Maybe a solution in future if CSS is a really big problem, but not worth it.

**Comment by bedeoverend** [comment]
Ok so another snag, `pre-wrap` works fine in `simpla-text`, but it doesn't in `simpla-article` because article includes a clearfix. The `content: '';` in `::after` rule seems to cause `pre-wrap` to add a hunk down beneath it.

**Comment by bedeoverend** [comment]
@seaneking could you pull down the [ff-nbsp](https://github.com/SimplaElements/simpla-richtext-behavior/tree/ff-nbsp) branch and give that a test run? It seems to fix it, barring when you repeatedly enter spaces, but can work on that - it's not yet the same as Chrome's behavior.

**Comment by madeleineostoja** [comment]
> So given that white-space: pre-wrap (or pre the only two that preserve whitespace) cause this nasty bug, we can't use a CSS fix - checked and it doesn't matter how it's applied.

What? Applying pre-wrap to the individual elements (ie `simpla-article p`, etc) fixes the issue perfectly on Firefox and doesn't cause any layout issues. Did you check that? The problem is only on hosts with block children. 

It shouldn't be applied to the host anyway, except for inline editors (where the host itself acts as a paragraph). That was just lazy CSSing on ProseMirror's part.

Hope you didn't spend much time on debugging this / the JS solution ha.

**Comment by bedeoverend** [comment]
Damn it - no I didn't realise that applying to just block children would fix. Should have taken more time in reading over your assessment. Alright I'll work on getting a stylesheet injection thing going tomorrow - could you list out the exact styles that'll need to be applied? Rather not attempt a generalised plugin to do this straight away.

**Comment by madeleineostoja** [comment]
Sure thang - for just this issue it'd be just `white-space: pre-wrap` on block leaf nodes (so just p and headings for now I think, since blockquote and list items wrap). 

```css
simpla-article p,
simpla-article h1,
simpla-article h2 {
  white-space: pre-wrap
}
```

Then a conditional to apply it to the host only if it's an inline editor (and not inject the other styles).

**Comment by madeleineostoja** [comment]
And in that case maybe open a new issue for conditional stylesheets based on plugins available (including placeholder).

We would want:
 - This hack if not inline - for p, and headings if available
 - Host hack if inline
 - List hack if lists available (until issue properly fixed)
 - Img styles from article if image available
 - Placeholder styles if placeholder set

And there's some other funky stuff in ProseMirror's default styles to do with lists, not sure what they're for though.

Also should move the issue on simpla-text about unecessary stylesheets here, to be looked at as an enhancement after ^ is done.

**Comment by bedeoverend** [comment]
So see attached PR. Worth noting that it appears firefox adds `&nbsp;` instead of regular spaces for inline editor i.e. when inserting directly into the host rather than children, so shouldn't need any hack for inline styles.

**Comment by madeleineostoja** [comment]
> Worth noting that it appears firefox adds &nbsp; instead of regular spaces for inline editor

Of course it does 🙄
